Grand Ritual: Alternate Wizard objective (Wizard Events II)

Jacquerel opened the pull request:
## About The Pull Request

Adds an alternate greentext objective for Wizards known as the "Grand Ritual". This was initially the gimmick of a different wizard-related antagonist downstream. I didn't get permission to port it, so I'm attaching it to regular Wizards instead.

![image](https://user-images.githubusercontent.com/7483112/214142793-a53c3399-d087-4b90-bf7c-a3fba9de5cf1.png)

Wizards will spawn in with a new Grand Ritual button next to their antagonist info button. Pressing it will pinpoint them towards their next Ritual Location (a randomly chosen region of the space station).
Once within that location, pressing it will summon a magic circle and obliterate any dense objects which are in the way. This also puts the ability on a two minute cooldown.
Clicking on the magic circle with an empty hand will begin a three-stage invocation to gather magical power. You can interrupt this invocation at any time and will resume from the last stage you completed (if you finished two stages you only need to do one more).
Once you complete a ritual, a random event will be triggered based on how many rituals you have performed so far. These tend to be ones which annoy the crew in some manner, and Wizard Events are included in the list. Additionally, something weird will usually happen to the room you are in.
Then you are assigned a new location and can toddle off to do it again.

Here's an example video of the process.
https://www.youtube.com/watch?v=iOT6iBFw0r0
You don't always scream between invocations that's just what it picked for this one.

Once you have done this three times, you will be picked up by the station's sensors every time you start a subsequent ritual and should expect annoyed company to come investigate.
Once you have done this six times, you can finally spend all of that accumulated power on the seventh Grand Finale ritual. Completing this grants you victory at the end of the round and will have a larger, flashier effect which you can pick from a list of options, think of it like a wizard equivalent of a Traitor Final Objective or Heretic Ascension.
After that you can still keep doing rituals if you want to pester the crew further by summoning more random events, you've already "won" at this point so now it's your job to make them want to go home.

I think it'd be more fun to just find out what the Finale ritual can do by seeing it happen but maintainers will probably want a list of its precise capabilities, so here it is:

<details>
  <summary>Click Me</summary>
You get to pick one of these options:

- Alter the past such that the Wizard is now the Captain (complete with Captain's ID and outfit) and any existing Captain is (has always been?) an Assistant. Gives an existing Captain and random other crew members objective to murder the wizard.
- Alter the past such that everyone on the space station has always also been a clown in addition to their regular job. Doles out clown suits which can produce random prank-related items once a minute. Gives pre-existing clowns the innate ability to do this regardless of costume. Makes Clown Enjoyers extremely happy, and makes a small percentage of other people gain objectives to murder the Wizard.
- Unbolt and then open every single door on the space station, then remove all access requirements from every single door on the space station.
- Transmute the entire space station slowly over time into a random choice of Gold, Pizza, Meat, or Wood. Technically also removes access restrictions from every door, but much more slowly.
- Cast Summon Magic, with a 20% chance per crewmember of gaining an objective to hoard artifacts.
- _If 90 minutes have passed_ perform a longer ritual which ends with the wizard gibbing themselves and summoning either a singularity, tesla ball, or endless meteor storm.

</details>

And for good measure, here's a list of random things which can happen to the room you are in when you finish a ritual:


<details>
  <summary>Click Me</summary>

- The room experiences "Space Time Distortion" in a similar manner to the spell effect, with a range based on the number of rituals completed so far.
- An area around the ritual circle is transmuted as if by a Dimensional Anomaly, with a ranged based on the number of rituals so far.
- A small number of less-dangerous anomalies temporarily appear in the area, quantity based on the number of rituals so far.
- A pulse of EMP in the area with a range based on... you probably get the idea at this point.
- Mobs near the area have their locations randomly shuffled.
- The nearby area is coated in lubricant.
- A random crew member anywhere on the station is momentarily paralysed, then summoned directly to the ritual location.
- A cloud of colourful smoke.
- Splatters the area in gore.
- Food begins to rain from the ceiling.
- A small number of hostile mobs appear with a short summoning delay.
- Some apes appear to serenade you with various instruments.

</details>

Currently completing a ritual also has a chance to create Heretic Reality Tears (of both varieties, available for Heretics to eat and visible to crew) as a kind of cross-antagonist interaction which seemed to make sense to me but if this seems thematically or mechanically inappropriate it's easy to strip out.

## Why It's Good For The Game

Wizard objectives, drawing from the pool of old-style traitor objectives, are simply extremely boring and have no flavour at all. Why would a wizard steal a sliver of supermatter anyway? This is both more unique and also more flavourful.
But if you're really attached to them (and they are likely easier) they're also still there. So you can still do that, and pop down a couple of runes as a distraction if you happen to be passing through.

Wizard _events_ are in a state where they have been soft-removed from the game by being limited to the extremely rare 100 threat round which _also_ happens to roll a Wizard. Not only does this mean the content has been unused, it also means a lot of it has rotted from disuse. This reintroduces some of the less immediately round-upending ones to the game in a fashion which avoids some of the pitfalls of the Wizard Events spellbook ritual. In this form the Wizard is always vulnerable while triggering the events, they're somewhat as effected by what they have done as you are, and you can stop the events from happening by killing them.

This is an objective which does not allow for long-term stealth as it leaves evidence behind and starts simply announcing your presence after a while. This encourages players not to be boring and spend a majority of the round's threat budget on someone who just jaunts into offices to steal items once they have confirmed as a ghost that there's nobody in there to see them.
It might even encourage use of some less used abilities, the period where you are vulnerable channeling your rune at least incentivises having minions around to protect you or some method of fortifying an area.

## Changelog
:cl:
add: Wizards can now perform a Grand Ritual to summon unpleasant events to the space station.
add: Wizards completing the Grand Ritual a sufficient number of times can achieve victory at the end of the round, and cast a powerful round-changing spell.
/:cl:

Jacquerel committed:
Adds grand ritual files, doesn't work yet because I forgot some

Jacquerel committed:
Ok almost done just missing icons

Jacquerel committed:
Adds icons, it compiles now

Jacquerel committed:
Correctly grant ritual

Jacquerel committed:
Fix some oversights
Grand Ritual has a 2 minute cooldown between runes
Don't transmute entire station into glass (RIP North Star)
Add grand ritual details to UI

Jacquerel committed:
Removes a literally empty file

Jacquerel committed:
Vitally important clown grammar update

Jacquerel committed:
Culls frankly excessive list of strings
Refactors some duplicated code

optimumtact commented on code/modules/antagonists/wizard/grand_ritual/grand_side_effect.dm:
```diff
@@ -0,0 +1,419 @@
+	qdel(lube)
+
+/// Grabs one person and pulls them to this location, after a delay
+/datum/grand_side_effect/summon_crewmate
```
how dare u

optimumtact commented on code/modules/clothing/masks/gasmask.dm:
```diff
@@ -1,3 +1,12 @@
+/// Map of names of clown mask types to clown mask icon states
+GLOBAL_LIST_INIT(clown_mask_options, list(
+	"True Form" = "clown",
+	"The Feminist" = "sexyclown",
```
this joke has worn out it's welcome and the name should be changed.

Time-Green commented:
This is so dumb and fun and I already love it so much

Jacquerel committed:
Replaces a bad old joke with, just a description.

Zytolg commented:
This is great

private-tristan commented:
Idk about the 90 minute events gibbing the wizard, seeing how a wizard would have likely killed a LOT of the crew by that point in the round, I don't think they'd be willing to die so that a skrung can spawn in and instantly wander into space (when they can just delam the SM and do the same thing without gibbing themselves).

how does it work with multiple wizards? do they all progress different rituals or do they all work together on the same ritual?

Jacquerel commented:
Progress is separate, so if there are several wizards they'll need to do seven each.

Shadowflame909 commented:
Seems difficult but cool. Perfect for the wizard!

SuperSlayer0 commented:
>A random crew member anywhere on the station is momentarily paralysed, then summoned directly to the ritual location.

Please remove this part, being randomly sent to imminent death only because some wizard at the other part of the station finished a ritual is extremly unfun.

ATH1909 commented:
> Unbolt and then open every single door on the space station, then remove all access requirements from every single door on the space station.

Does this bolt open doors that are adjacent to space, draining the station of air?

> Cast Summon Magic, with a 20% chance per crewmember of gaining an objective to hoard artifacts.

Could Summon Guns also be added to this list? ~~And/or an option to choose both?~~

Jacquerel commented:
It doesn’t bolt doors open, only removing the bolt from currently bolted doors. Doors to outside *will* open and vent some air but only for their natural cycle time. After that, anyone who wants to can just click them open though.  

I could add summon guns too but it feels a little redundant with the other option. 

I’ll think about the “grab someone over” effect because you’re probably right that they’ll just get killed on arrival, this was less of a problem for the antagonist i originally designed this for.

Helg2 commented:
Wont it be better if you add rituals not as a main target, but as free ritual in spellbook? Because i think if it'll be as main target it may happen that some people will just take blinks, teleports and etc. and wont do anything besides ritual in the round. And it'll be just loss of treat and major antag for the round.

Jacquerel commented:
If someone only takes teleport/transport spells then the crew will probably just kill them, which is indeed a major loss of threat but no worse than how people already have the ability to play wizard badly and die very fast in current rounds. 

I’m not sure this is more of an incentive to do that than existing theft objectives, this actively starts telling people where you are after the first few rituals and so not preparing to encounter the crew is a bit of an own goal.

Jacquerel committed:
Merge branch 'master' into wacky-wizard-wincon

Jacquerel committed:
Update for manifest changes

Jacquerel committed:
Merge branch 'master' into wacky-wizard-wincon

CCC23 commented:
finally biddle wizard

Redrover1760 commented:
> I’ll think about the “grab someone over” effect because you’re probably right that they’ll just get killed on arrival, this was less of a problem for the antagonist i originally designed this for.

I mean, if its a small enough chance to happen we can just chalk it up to "You should have went after the wizard." 

Or you could restrict the event until after the wizard's ritualing starts being announced.

Jacquerel committed:
Merge branch 'master' into wacky-wizard-wincon

Jacquerel committed:
I missed a conflict

Jacquerel committed:
Moves vars into correct class

Jacquerel committed:
Adds missing arg to override

san7890 commented on code/modules/antagonists/wizard/equipment/enchanted_clown_suit.dm:
```diff
@@ -0,0 +1,104 @@
+	var/casting = FALSE
+	/// List of prank implements you can find in your pockets
+	var/static/list/clown_items = list(
+		/obj/item/grown/bananapeel = 5,
+		/obj/item/food/pie/cream = 5,
+		/obj/item/bikehorn = 5,
```
alphabetize this entire list. there's a few VSC extensions that do this for you rather quickly and pain-free

san7890 commented on code/modules/antagonists/wizard/equipment/enchanted_clown_suit.dm:
```diff
@@ -0,0 +1,104 @@
+		/obj/item/bikehorn/golden = 1,
+		/obj/item/dualsaber/toy = 1,
+		/obj/item/toy/balloon/syndicate = 1,
+		/obj/item/toy/balloon/arrest = 1, )
```
```suggestion
		/obj/item/toy/balloon/arrest = 1,
	)
```

san7890 commented on code/modules/antagonists/wizard/grand_ritual/grand_ritual.dm:
```diff
@@ -0,0 +1,286 @@
+	var/datum/weakref/rune
+
+	/// A blacklist of turfs we cannot scribe on.
+	var/static/list/blacklisted_rune_turfs = typecacheof(list(/turf/closed/indestructible, /turf/open/indestructible, /turf/open/space, /turf/open/openspace, /turf/open/lava, /turf/open/chasm))
```
multi-line this list. i think the rule of thumb is three different things? unsure, but would benefit.

san7890 commented on code/modules/antagonists/wizard/grand_ritual/grand_ritual.dm:
```diff
@@ -0,0 +1,286 @@
+	 * To be honest if maintenance subtypes didn't exist I could probably have got away with just a blacklist, c'est la vie
+	 */
+	var/static/list/area_whitelist = typecacheof(list(
+		/area/station/commons,
```
alphabetize this list and the one below it. also the one i asked you to multi-line.

san7890 commented on code/modules/antagonists/wizard/grand_ritual/grand_rune.dm:
```diff
@@ -0,0 +1,335 @@
+	/// Magic words you say to invoke the ritual
+	var/list/magic_words = list()
+	/// Things you might yell when invoking a rune
+	var/static/list/possible_magic_words = list(
```
alphabetize. funny too

Jacquerel committed:
Alphabetises some lists.

Jacquerel committed:
Merge branch 'master' into wacky-wizard-wincon

Jacquerel committed:
Which asshole touched events? Oh... it was me.

Jacquerel committed:
Clown pockets extends conjure item
Don't pointlessly track a variable when we could just reset the cooldown when we set it to TRUE
Adds soul to balloon alert
Removes double parenthesis
Handle missing owner
Adds missing return
Create ritual action with owner when granting it
Don't copy list for no reason
Bug fix to a wizard event I triggered while testing
Whitespace/indenting of lists

Jacquerel committed:
The panel can be displayed before the ritual action is assigned.

Jacquerel committed:
Don't spawn meteors on icebox

Jacquerel committed:
What kind of idiot would make a PR which would conflict virtually any time anyone else touched an event

Jacquerel committed:
Missed a conflict marker

Jacquerel committed:
Merge branch 'master' into wacky-wizard-wincon

Jacquerel committed:
wat da hell icebox

MrMelbert reviewed (approved):
Minor nitpicks again, otherwise pretty much ready when these are done and tested

Jacquerel committed:
Add comment to unexploding flux anomaly
Create wizard ritual in `on_gain`, refactor it a little bit too
Check turf before allowing ritual drawing
Use helper procs to figure out what jobs people are
Announce when meteors are launched

Jacquerel commented on code/modules/antagonists/wizard/grand_ritual/grand_ritual.dm:
```diff
@@ -0,0 +1,296 @@
+	if (HAS_TRAIT(owner, TRAIT_MAGICALLY_PHASED))
+		if (feedback)
+			owner.balloon_alert(owner, "become tangible first!")
+		return FALSE
```
You can't now that I've made that change, which seems sensible.

Jacquerel commented on code/modules/antagonists/wizard/grand_ritual/grand_ritual_finale.dm:
```diff
@@ -0,0 +1,407 @@
+			var/obj/energy_ball/tesla = new (current_location)
+			tesla.energy = 200
+		if (DOOM_METEORS)
+			var/datum/dynamic_ruleset/roundstart/meteor/meteors = new()
+			meteors.meteordelay = 0
+			var/datum/game_mode/dynamic/mode = SSticker.mode
+			mode.execute_roundstart_rule(meteors) // Meteors will continue until morale is crushed.
```
Surprisingly yeah it does. I did realise when double checking that it doesn't give you any immediate feedback though so I made it print a message so that the player doesn't ahelp thinking that nothing happened.

Jacquerel commented on code/modules/antagonists/wizard/wizard.dm:
```diff
@@ -194,14 +205,28 @@ GLOBAL_LIST_EMPTY(wizard_spellbook_purchases_by_key)
+	var/mob/living/wizard_mob = mob_override || owner.current
+	wizard_mob.faction |= ROLE_WIZARD
+	add_team_hud(wizard_mob)
+	ritual = new(wizard_mob)
```
I have moved the instantiation to `on_gain`, applying it to the mob still needs to happen in `apply_innate_effects` or it is removed if you change mob and never comes back.

Jacquerel committed:
gotta stop conflicting my own PRs

Jacquerel committed:
Don't try to call removed method
Runes will be invisible to the AI while being drawn as well as when finished

Jacquerel committed:
get your organs out of there you dirty animal

Jacquerel committed:
Some of the review changes while I switch branch

Jacquerel commented on code/modules/antagonists/wizard/grand_ritual/grand_side_effect.dm:
```diff
@@ -0,0 +1,421 @@
+			break //locs aren't always unique, so this may come into play
+		do_teleport(victim, moblocs[moblocs.len], channel = TELEPORT_CHANNEL_MAGIC)
+		moblocs.len -= 1
+	for (var/mob/living/victim in GLOB.alive_mob_list)
```
Using this list here doesn't even make logical sense I don't know why I did that

Jacquerel committed:
Refactor some loops and order of operations
Don't randomly spawn smoke on every mob on the map (??)
Don't apply most effects to nuclear operatives or ninjas
Other minor review notes

Jacquerel committed:
Moves CHECK_TICK to correct proc

Jacquerel committed:
More lists I missed

Jacquerel commented on code/modules/antagonists/wizard/grand_ritual/grand_side_effect.dm:
```diff
@@ -0,0 +1,421 @@
+		nearby_turfs += potential_turf
+
+	var/turf/landing_pos = pick(nearby_turfs)
+	var/obj/effect/temp_visual/teleport_abductor/landing_indicator = new(landing_pos)
+	landing_indicator.duration = CREWMATE_SUMMON_TELEPORT_DELAY
```
Turns out it's 9 seconds vs 8 seconds so I am not sure why I bothered in the first place

Jacquerel committed:
Remove line which didn't do anything

Jacquerel committed:
Oh yeah this doesn't need to be a var any more

Jacquerel committed:
Fixes order of operations for clown world such that it's valid for a player monkey to be given an objective to kill a wizard
Adds new proc and macro which allow me to search for 7 antag datums at once in a few places

Jacquerel commented on code/modules/antagonists/wizard/grand_ritual/grand_side_effect.dm:
```diff
@@ -0,0 +1,421 @@
+	if (ismonkey(crewmate))
+		return FALSE
+	if (!crewmate.mind)
+		return FALSE
```
The nuke op oversight really sent me down a rabbit hole let me know if I overengineered this

Jacquerel commented on code/__DEFINES/antagonists.dm:
```diff
@@ -229,6 +235,20 @@ GLOBAL_LIST_INIT(ai_employers, list(
+))
+
+/// Returns true if the given mob has an antag datum which is assigned to a human antagonist who doesn't spawn on the space station
+#define IS_HUMAN_INVADER(mob) (mob?.mind?.has_antag_datum_in_list(GLOB.human_invader_antagonists))
```
Pointing out that I could target nuke ops with some of these effects really sent me down a rabbit hole, let me know if you think I overengineered this as a solution

Jacquerel committed:
Merge branch 'master' into wacky-wizard-wincon

Jacquerel committed:
Adds ritual potency for the funny ghost anomaly

MrMelbert reviewed (approved):
I can nitpick a small few things still but eh, we ball

TheSilverNuke commented:
Can we update the wiki with this information? None of it is listed on there and is now out of date

Jacquerel commented:
> Can we update the wiki with this information? None of it is listed on there and is now out of date

Done 👍
